Src/Fido2/AuthenticatorAssertionResponse.cs

```diff
@@ -167,16 +182,210 @@ public static AuthenticatorAssertionResponse Parse(AuthenticatorAssertionRawResp
+                            return devicePublicKeyAuthenticatorOutput.GetBytes();
+                        }
+                        catch
+                        {
+                            // Some form of error has occurred. It is indeterminate whether this is a known device. Terminate these verification steps.
+                            throw;
+                        }
+                    }
+                }
+                else
+                {
+                    // Otherwise there is some form of error: we recieved a known dpk value, but one or more of the
+                    // accompanying aaguid, scope, or fmt values did not match what the Relying Party has stored
+                    // along with that dpk value. Terminate these verification steps.
+                    throw new Fido2VerificationException(Fido2ErrorCode.MissingStoredPublicKey, Fido2ErrorMessages.MissingStoredPublicKey);
+                }
+            }
+        }
+
+        // Otherwise, the Relying Party does not have attObjForDevicePublicKey fields presently mapped to this user account and credential.id pair:
+        else
+        {
+            // If fmt’s value is "none" there is no attestation signature to verify.
+            // Complete the steps in § 7.2 Verifying an Authentication Assertion and, if those steps are successful, store the extracted aaguid, dpk, scope, fmt, attStmt values indexed to the credential.id in the user account.
+            // Terminate these verification steps.
+            if (devicePublicKeyAuthenticatorOutput.Fmt.Equals("none"))
+                return devicePublicKeyAuthenticatorOutput.GetBytes();
+            // Otherwise, verify that attStmt is a correct attestation statement, conveying a valid attestation signature, by using the attestation statement format fmt’s verification procedure given attStmt. See § 10.2.2.2.2 Attestation calculations.
+            // Relying Party policy may specifiy which attestations are acceptable.
+            else
+            {
+                var verifier = AttestationVerifier.Create(devicePublicKeyAuthenticatorOutput.Fmt);
+                // https://w3c.github.io/webauthn/#sctn-device-publickey-attestation-calculations
+                try
+                {
+                    // This is a known device public key with a valid signature and valid attestation and thus a known device. Terminate these verification steps.
+                    _ = verifier.Verify(devicePublicKeyAuthenticatorOutput.AttStmt, devicePublicKeyAuthenticatorOutput.AuthData, devicePublicKeyAuthenticatorOutput.Hash);
+                    return devicePublicKeyAuthenticatorOutput.GetBytes();
+                }
+                catch
+                {
+                    // Some form of error has occurred. It is indeterminate whether this is a known device. Terminate these verification steps.
```

Comment:
Throw FidoVerificationException with inner exception?

Reply:
Yeah, I meant to go back and figure out what to throw for each of these but hadn't yet...and now I have to revisit again as some things have changed.